overlord/snapstate/snapmgr.go

```diff
@@ -386,6 +418,96 @@ func (m *SnapManager) blockedTask(cand *state.Task, running []*state.Task) bool
 	return false
 }
 
+var CanAutoRefresh func(st *state.State) bool
+
+// ensureRefreshes ensures that we refresh all installed snaps periodically
+func (m *SnapManager) ensureRefreshes() error {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	// see if it even makes sense to try to refresh
+	if CanAutoRefresh == nil || !CanAutoRefresh(m.state) {
+		return nil
+	}
+
+	tr := config.NewTransaction(m.state)
+
+	// for tests
+	var refreshDisabled bool
+	err := tr.Get("core", "refresh.disabled", &refreshDisabled)
+	if err != nil && !config.IsNoOption(err) {
+		return err
+	}
+	if refreshDisabled {
+		return nil
+	}
+
+	var lastRefresh time.Time
+	err = tr.Get("core", "refresh.last", &lastRefresh)
+	if err != nil && !config.IsNoOption(err) {
+		return err
+	}
+
+	var refreshRand time.Duration
+	err = tr.Get("core", "refresh.rand", &refreshRand)
+	if err != nil && !config.IsNoOption(err) {
+		return err
+	}
+
+	nextRefresh := lastRefresh.Add(minRefreshInterval).Add(refreshRand)
+	if time.Now().Before(nextRefresh) {
+		return nil
+	}
+
+	// check that there is no change in flight already
+	for _, chg := range m.state.Changes() {
+		if chg.Kind() == "auto-refresh" && !chg.Status().Ready() {
+			// change already in motion
+			return nil
+		}
+	}
+
+	// For setLastRefresh() we have two options here:
+	//  1. setLastRefresh only if no error happened
+	//     below in the "AutoRefresh()" call
+	//  2. always set the last refresh regardless of errors
+	//     in "AutoRefresh()"
+	//
+	// This currently implements (2) because if there are e.g. store
```

Comment:
Using (2) sounds a bit problematic now that we're giving people the option to dial down the frequency of retries. If someone has it set for once a week on Mondays, not retrying means it'll take another week until it tries again. Given how often we've been observing various kind of spurious errors, we should probably allow it to actually retry.

It should be super cheap for the store to say "go away, I have nothing for you!" if it really wants the retry to not happen. It can't easily say "Oh, come back please!", though.